File: java/src/main/java/abstract_factory/product/UltraBus.java
package abstract_factory.product;

import abstract_factory.product.part.VehiclePartFactory;

public class UltraBus extends Vehicle {
    public static final String NAME = "Ultra Bus";
    private final VehiclePartFactory vehiclePartFactory;

    public UltraBus(VehiclePartFactory vehiclePartFactory) {
        name = NAME;
        this.vehiclePartFactory = vehiclePartFactory;
    }

    @Override
    public void assembly() {
        engine = vehiclePartFactory.createEngine();
        tire = vehiclePartFactory.createTire();
        options = vehiclePartFactory.createOptions();
    }
}

File: kotlin/src/main/kotlin/abstract_factory/product/UltraBus.kt
package abstract_factory.product

import abstract_factory.product.part.VehiclePartFactory

class UltraBus(private val vehiclePartFactory: VehiclePartFactory) : Vehicle() {
    override var name = NAME

    override fun assembly() {
        engine = vehiclePartFactory.createEngine()
        tire = vehiclePartFactory.createTire()
        options = vehiclePartFactory.createOptions()
    }
    
    companion object {
        const val NAME: String = "Ultra Bus"
    }
}